prov/sm2/src/sm2_init.c

```diff
@@ -0,0 +1,186 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <rdma/fi_errno.h>
+
+#include "sm2.h"
+#include "sm2_signal.h"
+#include <ofi_hmem.h>
+#include <ofi_prov.h>
+
+struct sigaction *sm2_old_action = NULL;
+
+/*
+ * Convert strings node + service into a single string addr
+ */
+static void sm2_resolve_addr(const char *node, const char *service, char **addr,
+			     size_t *addrlen)
+{
+	char temp_name[SM2_NAME_MAX];
+
+	FI_INFO(&sm2_prov, FI_LOG_EP_CTRL, "resolving node=%s, service=%s\n",
+		node ? node : "NULL", service ? service : "NULL");
+	if (service) {
+		if (node)
+			*addrlen =
+				snprintf(temp_name, SM2_NAME_MAX - 1, "%s%s:%s",
+					 SM2_PREFIX_NS, node, service);
+		else
+			*addrlen = snprintf(temp_name, SM2_NAME_MAX - 1, "%s%s",
+					    SM2_PREFIX_NS, service);
+	} else {
+		if (node)
+			*addrlen = snprintf(temp_name, SM2_NAME_MAX - 1, "%s%s",
+					    SM2_PREFIX, node);
+		else
+			*addrlen = snprintf(temp_name, SM2_NAME_MAX - 1, "%s%d",
+					    SM2_PREFIX, getpid());
+	}
+	*addr = strndup(temp_name, SM2_NAME_MAX - 1);
+	FI_INFO(&sm2_prov, FI_LOG_EP_CTRL, "resolved to %s\n", temp_name);
+}
+
+/*
+ * The sm2_shm_space_check is to check if there's enough shm space we
+ * need under /dev/shm.
+ * Here we use #core instead of sm2_MAX_PEERS, as it is the most likely
+ * value and has less possibility of failing fi_getinfo calls that are
+ * currently passing, and breaking currently working app
+ */
+static int sm2_shm_space_check(size_t tx_count, size_t rx_count)
+{
+	/* TODO: call AFTER we have mmap, but BEFORE we allocate space. */
+	/* TODO: Base return value on the contents of header region size. */
+	/* TODO: ignore existing file allocation size when stat() /dev/shm/ */
+	struct statvfs stat;
+	char shm_fs[] = "/dev/shm";
+	uint64_t available_size, shm_size_needed;
+	int num_of_core, err;
+
+	num_of_core = ofi_sysconf(_SC_NPROCESSORS_ONLN);
+	if (num_of_core < 0) {
+		FI_WARN(&sm2_prov, FI_LOG_CORE,
+			"Get number of processors failed (%s)\n",
+			strerror(errno));
+		return -errno;
+	}
+	shm_size_needed = num_of_core * sm2_calculate_size_offsets(NULL, NULL);
+	err = statvfs(shm_fs, &stat);
+	if (err) {
+		FI_WARN(&sm2_prov, FI_LOG_CORE,
+			"Get filesystem %s statistics failed (%s)\n", shm_fs,
+			strerror(errno));
+	} else {
+		available_size = stat.f_bsize * stat.f_bavail;
+		if (available_size < shm_size_needed) {
+			FI_WARN(&sm2_prov, FI_LOG_CORE,
+				"Not enough available space in %s.\n", shm_fs);
+			return -FI_ENOSPC;
+		}
+	}
+	/* TODO: we should ignore space already reserved in the global shm file.
+	 */
```

Comment:
I don't think I understand this TODO?

Reply:
The "is there enough space left" check is looking to see if the remaining space left on the system is enough for our SHM allocation, BUT we already have a big file that might have an entry in it... i.e.. if the SHM file is already 4096, why do we need to check for more space.